### PR TITLE
Taproot :)

### PR DESCRIPTION
**Pull request by elichai**
Sorry. was opened in mistake.

### CODE REVIEW

File: src/script/script.h [hunk 1]
```diff
@@ -583,4 +595,15 @@ struct CScriptWitness
     std::string ToString() const;
 };
 
+class CReserveScript
```
Comment (jnewbery) [inline src/script/script.h]: You've accidentally reintroduced this in ebcfe3584199f281f825625dcadef8535dece536

File: src/script/interpreter.cpp [hunk 2]
```diff
@@ -908,54 +911,105 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
...
+                        execdata.m_scriptcode = CScript(pbegincodehash, pend);
 
-                    if (!CheckSignatureEncoding(vchSig, flags, serror) || !CheckPubKeyEncoding(vchPubKey, flags, sigversion, serror)) {
-                        //serror is set
-                        return false;
-                    }
-                    bool fSuccess = checker.CheckSig(vchSig, vchPubKey, scriptCode, sigversion);
+                        // Drop the signature in pre-segwit scripts but not segwit scripts
+                        if (sigversion == SigVersion::BASE) {
+                            int found = FindAndDelete(execdata.m_scriptcode,  CScript() << vchSig);
```
Comment (jnewbery) [inline src/script/interpreter.cpp]: nit: double space before `CScript`

File: src/validation.cpp [hunk 3]
```diff
@@ -1689,6 +1695,11 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
         flags |= SCRIPT_VERIFY_CHECKSEQUENCEVERIFY;
     }
 
+    // Start enforcing Taproot using versionbits logic.
+    if (VersionBitsState(pindex->pprev, consensusparams, Consensus::DEPLOYMENT_TAPROOT, versionbitscache) == ThresholdState::ACTIVE) {
+        flags |= SCRIPT_VERIFY_TAPROOT;
+    }
+
```
Comment (jnewbery) [inline src/validation.cpp]: nit: trailing whitespace

**Comment by jnewbery** [comment]
opened in error. Should have been on a different repo

File: src/script/script.h (hunk 1 above)
Comment (elichai) [reply src/script/script.h]: What do you mean? isn't https://github.com/bitcoin/bitcoin/commit/ebcfe3584199f281f825625dcadef8535dece536 the only thing where this exists?

**Pull request closed without merging**